src/rust/engine/fs/src/store.rs

```diff
@@ -275,8 +275,12 @@ mod local {
 
   impl ByteStore {
     pub fn new<P: AsRef<Path>>(path: P, pool: Arc<ResettablePool>) -> Result<ByteStore, String> {
-      // 2 DBs; one for file contents, one for directories.
       let env = Environment::new()
+        // Without this flag, read transactions don't give up their reader slots until the thread
```

Comment:
Was this detected via benchmarking? Thread local storage sounds... like a pretty good thing in general.

Reply:
Without this, our own test suite fails hard by exhausting the transaction count.

I don't think it's anywhere near as scary as its name suggests :)

I believe the mechanism here is that there is a finite number of transaction slots (default: 126), and without this flag, each time an OS thread opens a transaction, it takes up a slot, and those slots only get freed up when the OS thread dies. If you make multiple transactions sequentially from the same thread (say, you're using a thread pool), you accordingly exhaust your slots after 126 transactions, and can never read from the DB while those threads are alive.

The only down-side is that you need to make sure that any individual thread must not try to perform multiple write transactions concurrently. We don't try to run multiple write transactions concurrently on any particular OS thread, so we're fine there. (At least, that's my understanding from https://docs.rs/lmdb/0.7.0/lmdb/constant.NO_TLS.html - I guess the other interpretation is that we need a Mutex around attempts to create new write transactions, but that's not how I read it).

The Go and Python bindings turn this option on by default. It definitely feels like this should be the default behaviour in general...

Reply:
Great, thanks. Worth expanding the comment a tiny bit probably, as I expected that it was a performance tunable.